diff --git a/include/linux/sched.h b/include/linux/sched.h
index 3dd0efb..937ab61 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1804,6 +1804,9 @@ static inline void put_task_struct(struct task_struct *t)
 extern void task_times(struct task_struct *p, cputime_t *ut, cputime_t *st);
 extern void thread_group_times(struct task_struct *p, cputime_t *ut, cputime_t *st);
 
+extern int task_free_register(struct notifier_block *n);
+extern int task_free_unregister(struct notifier_block *n);
+
 /*
  * Per process flags
  */
