diff --git a/include/linux/cpu.h b/include/linux/cpu.h
index 78ed62f..caec85b 100644
--- a/include/linux/cpu.h
+++ b/include/linux/cpu.h
@@ -213,4 +213,11 @@ static inline int disable_nonboot_cpus(void) { return 0; }
 static inline void enable_nonboot_cpus(void) {}
 #endif /* !CONFIG_PM_SLEEP_SMP */
 
+#define IDLE_START 1
+#define IDLE_END 2
+
+void idle_notifier_register(struct notifier_block *n);
+void idle_notifier_unregister(struct notifier_block *n);
+void idle_notifier_call_chain(unsigned long val);
+
 #endif /* _LINUX_CPU_H_ */
