diff --git a/include/linux/videodev2.h b/include/linux/videodev2.h
index c9c9a46..6490191 100644
--- a/include/linux/videodev2.h
+++ b/include/linux/videodev2.h
@@ -1208,6 +1208,14 @@ struct v4l2_querymenu {
 #define V4L2_CID_BLACK_LEVEL		(V4L2_CID_BASE+11) /* Deprecated */
 #define V4L2_CID_AUTO_WHITE_BALANCE	(V4L2_CID_BASE+12)
 #define V4L2_CID_DO_WHITE_BALANCE	(V4L2_CID_BASE+13)
+enum v4l2_whiteblance {
+	V4L2_WB_AUTO = 0,
+	V4L2_WB_CLOUD = 1,
+	V4L2_WB_DAYLIGHT = 2,
+	V4L2_WB_INCANDESCENCE = 3,
+	V4L2_WB_FLUORESCENT = 4,
+	V4L2_WB_TUNGSTEN = 5,
+};
 #define V4L2_CID_RED_BALANCE		(V4L2_CID_BASE+14)
 #define V4L2_CID_BLUE_BALANCE		(V4L2_CID_BASE+15)
 #define V4L2_CID_GAMMA			(V4L2_CID_BASE+16)
@@ -1266,6 +1274,15 @@ enum v4l2_colorfx {
 
 /* last CID + 1 */
 #define V4L2_CID_LASTP1                         (V4L2_CID_BASE+42)
+/* Add camera flash light by raymonxiu */
+#define V4L2_CID_CAMERA_FLASH_MODE              (V4L2_CID_PRIVATE_BASE+0)
+enum v4l2_flash_mode {
+	V4L2_FLASH_MODE_OFF = 0,
+	V4L2_FLASH_MODE_AUTO = 1,
+	V4L2_FLASH_MODE_ON = 2,
+	V4L2_FLASH_MODE_TORCH = 3,
+	V4L2_FLASH_MODE_RED_EYE = 4,
+};
 
 /*  MPEG-class control IDs defined by V4L2 */
 #define V4L2_CID_MPEG_BASE 			(V4L2_CTRL_CLASS_MPEG | 0x900)
